hazelcast/sql.py

```diff
@@ -0,0 +1,1629 @@
+        self._sql_service = sql_service
+        """_InternalSqlService: Reference to the SQL service."""
+
+        self._connection = connection
+        """hazelcast.connection.Connection: Reference to the connection
+        that the execute request is made to."""
+
+        self._query_id = query_id
+        """_SqlQueryId: Unique id of the SQL query."""
+
+        self._cursor_buffer_size = cursor_buffer_size
+        """int: Size of the cursor buffer measured in the number of rows."""
+
+        self._lock = RLock()
+        """RLock: Protects the shared access to instance variables below."""
+
+        self._execute_response = Future()
+        """Future: Will be resolved with :class:`_ExecuteResponse` once the
+        execute request is resolved."""
+
+        self._iterator_requested = False
+        """bool: Flag that shows whether an iterator is already requested."""
+
+        self._closed = False
+        """bool: Flag that shows whether the query execution is still active
+        on the server side. When ``True``, there is no need to send the "close" 
+        request to the server."""
+
+        self._fetch_future = None
+        """Future: Will be set, if there are more pages to fetch on the server
+        side. It should be set to ``None`` once the fetch is completed."""
+
+        execute_future.add_done_callback(self._handle_execute_response)
+
+    def iterator(self):
+        """Returns the iterator over the result rows.
+
+        The iterator may be requested only once.
+
+        Returns:
+            Future[Iterator[Future[SqlRow]]]: Iterator that produces Future
+            of :class:`SqlRow` s. See the class documentation for the correct
+            way to use this.
+        """
+        return self._get_iterator(False)
+
+    def is_row_set(self):
+        """
+        Returns:
+            Future[bool]: Whether this result has rows to iterate.
+        """
+
+        def continuation(future):
+            response = future.result()
+            # By design, if the row_metadata (or row_page) is None,
+            # we only got the update count.
+            return response.row_metadata is not None
+
+        return self._execute_response.continue_with(continuation)
+
+    def update_count(self):
+        """Returns the number of rows updated by the statement or ``-1`` if this
+        result is a row set. In case the result doesn't contain rows but the
+        update count isn't applicable or known, ``0`` is returned.
+
+        Returns:
+            Future[int]:
+        """
+
+        def continuation(future):
+            response = future.result()
+            # This will be set to -1, when we got row set on the client side.
+            # See _on_execute_response.
+            return response.update_count
+
+        return self._execute_response.continue_with(continuation)
+
+    def get_row_metadata(self):
+        """Gets the row metadata.
+
```

Comment:
Raises:

HazelcastSqlError: In case of an SQL execution error.
ValueError: If the result contains only update count.

(Note: raises may not be correct since future is rejected with these errors, you can find the correct wording)